### ISSUE TITLE
[RFC]: FlashMask Attention Backend for PrefixLM Models

### ISSUE THREAD

**Issue by lucianommartins** [vllm-project/vllm#30786]
### Motivation.

PrefixLM models like Gemma3 multimodal require mixed attention patterns: bidirectional attention for image tokens (prefix) combined with causal attention for text tokens. Currently, vLLM handles this pattern using TritonAttention (or FlexAttention as fallback), which is slower than FlashAttention-2.

The core issue is that FlashAttention-2 only supports pure causal or pure non-causal attention—it cannot represent the hybrid mask pattern:

```
Image Tokens [0, N):      Bidirectional (attend to all image tokens)
Text Tokens [N, seq_len): Causal + attend to prefix
```

This forces PrefixLM models to use slower backends, creating a significant performance penalty for multimodal inference.

From benchmarks on A100-80GB:

| Backend | gemma3-4b seq=512 | gemma3-27b seq=512 | Notes |
|---------|-------------------|---------------------|-------|
| **FlashAttention-2** | 0.072 ms | 0.082 ms | Text-only baseline |
| **TritonAttention** | 0.18 ms | 0.16 ms | Current PrefixLM path |
| **FlexAttention** | 0.29 ms | 0.33 ms | Fallback |
| **Ratio** | 2.5x slower | 2x slower | Overhead vs FlashAttn |

The 2-4x overhead applies to every attention layer, accumulating to significant end-to-end latency.

Models with `is_mm_prefix_lm=True` that would benefit:
- Gemma3 (4B/12B/27B) - Multimodal with bidirectional image prefix
- Future multimodal models - Any model requiring partial bidirectional attention
- PrefixLM architectures - T5-style prefix attention patterns

Strategic value:

1. Significant performance improvement for multimodal: Closing the gap with FlashAttention
2. O(N) memory for attention masks vs O(N²): critical for long context
3. Unified backend: Single backend for text and multimodal (simpler codebase)
4. Future-proof: Extensible to other complex mask patterns

### Proposed Change.

FlashMask (ICLR 2025 Accepted, Wang et al.) extends FlashAttention-2 with column-wise sparse mask representation, enabling arbitrary attention patterns with O(N) memory complexity.

Paper: "FlashMask: Efficient and Rich Mask Extension of FlashAttention"
Status: ICLR 2025 Poster
Authors: Guoxia Wang, Jinle Zeng, Xiyuan Xiao, Siming Wu, Jiabin Yang, et al.

Column-Wise Sparse Mask Representation

Instead of storing a dense O(N²) mask, FlashMask represents attention patterns as column indices per row:

```
# For each query row i, store the start/end columns that can be attended to
mask_representation = {
    "col_left": [left_bound_0, left_bound_1, ...],   # [seq_len]
    "col_right": [right_bound_0, right_bound_1, ...]  # [seq_len]
}

# Attention allowed: col_left[i] <= j <= col_right[i]
```

For Gemma3 PrefixLM pattern with 256 image tokens:

```
# Image tokens (rows 0-255): attend to all image tokens
col_left[0:256] = 0
col_right[0:256] = 255

# Text tokens (rows 256+): causal + attend to all prefix
col_left[256:] = 0      # Can attend back to token 0 (prefix)
col_right[i] = i         # Causal: attend up to current position
```

This representation is O(seq_len) instead of O(seq_len²).

The FlashMask kernel extends FlashAttention-2's tiling algorithm:

```cpp
// FlashAttention-2 base (simplified)
for (int k_block = 0; k_block < num_k_blocks; k_block++) {
    // Load Q, K, V tiles
    // Compute attention scores
    for (int j = 0; j < BLOCK_N; j++) {
        // Causal mask: if (col > row) mask = -inf
        bool valid = (k_block * BLOCK_N + j) <= query_row;
        scores[j] = valid ? qk_dot[j] : -INFINITY;
    }
    // Online softmax, accumulate output
}

// FlashMask extension
for (int k_block = 0; k_block < num_k_blocks; k_block++) {
    int col_start = k_block * BLOCK_N;
    
    // Load column bounds for current query rows
    int left = col_left[query_row];
    int right = col_right[query_row];
    
    // Column-wise sparse mask
    for (int j = 0; j < BLOCK_N; j++) {
        int col = col_start + j;
        // Check against column bounds (replaces causal check)
        bool valid = (col >= left) && (col <= right);
        scores[j] = valid ? qk_dot[j] : -INFINITY;
    }
    // Rest unchanged: online softmax, accumulate output
}
```

The mask check remains O(1) per element—only the comparison changes.

Proposed implementation plan

Phase 1: FlashMask Kernel Integration

Implement FlashMask CUDA kernels based on the paper algorithm:

```
Target: vllm/attention/ops/flashmask/

Files to create:
├── flashmask_fwd.cu          # Forward pass kernel
├── flashmask_bwd.cu          # Backward pass (if needed for training)
└── flashmask_utils.cuh       # Shared utilities
```

Modifications required:
- Adapt to vLLM's tensor layouts and dtypes
- Add paged attention block table support
- Integrate with vLLM's CUDA graph capture

Python wrapper:

```
# vllm/attention/ops/flashmask.py

def flashmask_attention(
    query: torch.Tensor,      # [batch, seq_q, num_heads, head_dim]
    key: torch.Tensor,        # [batch, seq_k, num_kv_heads, head_dim]  
    value: torch.Tensor,      # [batch, seq_k, num_kv_heads, head_dim]
    col_left: torch.Tensor,   # [batch, seq_q] - left column bounds
    col_right: torch.Tensor,  # [batch, seq_q] - right column bounds
    scale: Optional[float] = None,
    window_size: Optional[int] = None,  # For sliding window layers
) -> torch.Tensor:
    """FlashMask attention with column-wise sparse masking."""
    ...
```

Phase 2: vLLM Backend Integration

2.1 Backend registration

```
# vllm/attention/backends/flashmask.py

class FlashMaskBackend(AttentionBackend):
    """FlashMask attention backend for PrefixLM models."""
    
    @staticmethod
    def get_name() -> str:
        return "FLASHMASK"
    
    @staticmethod
    def get_impl_cls() -> type["FlashMaskImpl"]:
        return FlashMaskImpl
    
    @classmethod
    def supports_mm_prefix(cls) -> bool:
        return True  # Critical: enables selection for PrefixLM
    
    @staticmethod
    def get_supported_head_sizes() -> list[int]:
        return [64, 96, 128, 192, 256]  # Include Gemma3 variants
    
    @staticmethod
    def get_min_capability() -> int:
        return 80  # Ampere minimum (same as FlashAttention-2)
```

2.2 Metadata extension

```
# Extend attention metadata for FlashMask

@dataclass
class FlashMaskMetadata(AttentionMetadata):
    # Column bounds for FlashMask sparse representation
    col_left: Optional[torch.Tensor] = None   # [total_tokens]
    col_right: Optional[torch.Tensor] = None  # [total_tokens]
    
    # Per-layer sliding window (for Gemma3's 5:1 pattern)
    layer_sliding_window: Optional[int] = None
```

2.3 Mask construction

```
# vllm/v1/worker/gpu_model_runner.py

def _build_flashmask_metadata(
    self,
    seq_lens: List[int],
    mm_prefix_ranges: List[Optional[Tuple[int, int]]],
    sliding_window: Optional[int],
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Build FlashMask column bounds from mm_prefix_range.
    
    For each sequence with prefix range [start, end):
    - Rows in [start, end): col_left=start, col_right=end-1 (bidirectional)
    - Rows >= end: col_left=0, col_right=current_row (causal + prefix)
    """
    col_left_list = []
    col_right_list = []
    
    for seq_len, mm_range in zip(seq_lens, mm_prefix_ranges):
        if mm_range is None:
            # Pure causal (no prefix)
            col_left_list.append(torch.zeros(seq_len, dtype=torch.int32))
            col_right_list.append(torch.arange(seq_len, dtype=torch.int32))
        else:
            prefix_start, prefix_end = mm_range
            col_left = torch.zeros(seq_len, dtype=torch.int32)
            col_right = torch.arange(seq_len, dtype=torch.int32)
            
            # Bidirectional within prefix region
            col_right[:prefix_end] = prefix_end - 1
            
            col_left_list.append(col_left)
            col_right_list.append(col_right)
    
    return torch.cat(col_left_list), torch.cat(col_right_list)
```

2.4 Backend selection update

```
# vllm/attention/selector.py

def get_attention_backend(model_config, ...):
    if model_config.is_mm_prefix_lm:
        # Priority order for PrefixLM models
        if FlashMaskBackend.is_available():
            return FlashMaskBackend  # NEW: Preferred
        elif TritonAttentionBackend.is_available():
            return TritonAttentionBackend  # Fallback
        else:
            return FlexAttentionBackend  # Last resort
    # ... rest unchanged
```

Phase 3: Testing & Validation

3.1 Correctness tests

```
# tests/attention/test_flashmask.py

def test_flashmask_causal_matches_flashattn():
    """Pure causal should match FlashAttention exactly."""
    
def test_flashmask_prefixlm_matches_tritonattn():
    """PrefixLM pattern should match TritonAttention output."""
    
def test_flashmask_gemma3_multimodal():
    """End-to-end Gemma3 multimodal produces correct output."""
    
def test_flashmask_sliding_window():
    """Sliding window + PrefixLM works correctly."""
    
def test_flashmask_paged_attention():
    """Works with PagedAttention block tables."""
```

3.2 Performance benchmarks

```
# benchmarks/attention/benchmark_flashmask.py

configs = [
    ("gemma3-4b", "text-only", [128, 512, 2048, 8192]),
    ("gemma3-4b", "multimodal-256", [128, 512, 2048]),
    ("gemma3-27b", "text-only", [128, 512, 2048, 8192]),
    ("gemma3-27b", "multimodal-256", [128, 512, 2048]),
]

# Targets:
# - Text-only: Close to FlashAttention-2
# - Multimodal: Significant improvement over TritonAttention
```

Hardware support

| GPU | SM Version | FlashMask Support | Fallback |
|-----|------------|-------------------|----------|
| H100 | SM 9.0 | Full | N/A |
| A100 | SM 8.0 | Full | N/A |
| RTX 4090 | SM 8.9 | Full | N/A |
| RTX 3090 | SM 8.6 | Full | N/A |
| V100 | SM 7.0 | No | TritonAttention |

FlashMask requires SM 8.0+ (same as FlashAttention-2).

File changes summary

| Path | Change | Description |
|------|--------|-------------|
| `vllm/attention/backends/flashmask.py` | NEW | Backend registration |
| `vllm/attention/backends/flashmask_impl.py` | NEW | Implementation |
| `vllm/attention/ops/flashmask.py` | NEW | Python wrapper |
| `csrc/flashmask/` | NEW | CUDA kernels |
| `vllm/attention/selector.py` | MODIFY | Add FlashMask selection |
| `vllm/v1/worker/gpu_model_runner.py` | MODIFY | Build FlashMask metadata |
| `tests/attention/test_flashmask.py` | NEW | Correctness tests |
| `benchmarks/attention/benchmark_flashmask.py` | NEW | Performance benchmarks |

Alternatives considered

1. Continue using TritonAttention

Status: Current solution
Pros: Already implemented, correct behavior
Cons: 2-4x slower than FlashAttention, not optimal for production

2. FlexAttention

Status: Available as fallback
Pros: Supports arbitrary masks
Cons: slower than FlashAttention, high compile overhead

3. Fused QKNormRoPE optimization

Status: Investigated, not viable for Gemma3
Reason: GemmaRMSNorm uses `x*(1+w)` formula, incompatible with existing fused kernel
Impact: Would only provide small performance gains even if implemented

4. Hybrid backend runtime switching

Status: Not possible in current vLLM architecture
Reason: vLLM selects attention backend once at model initialization
Would require: major architectural changes to support per-request backend selection

### Feedback Period.

2 weeks (14 days) from submission date.

### CC List.

- @WoosukKwon (vLLM maintainer, attention backends)
- @ywang96 (vLLM maintainer, attention backends)
- @zhuohan123 (vLLM maintainer, PagedAttention author)
- @youkaichao (vLLM committer, model integration)
- @DarkLight1337 (vLLM committer, multimodal models)
- @Isotr0py (vLLM committer, attention backends)
- @mgoin (vLLM committer, performance optimization)

### Any Other Things.

Implementation Commitment

I am prepared to contribute the implementation with maintainer guidance.

Risk Mitigation

1. Correctness: Extensive numerical testing against TritonAttention reference
2. CUDA Graph compatibility: Test with PIECEWISE mode early
3. Fallback: Keep TritonAttention as fallback for unsupported GPUs

Related Work

- PR #30386: Already added TritonAttention PrefixLM support
- FlexAttention: torch.compile-based solution
- This RFC: Natural evolution toward improved multimodal performance

Success Criteria

Correctness (Required):
1. FlashMask output matches TritonAttention output within numerical tolerance
2. All existing tests pass: `tests/models/decoder_only/vision_language/test_gemma3.py`
3. GGUF multimodal tests pass: `tests/models/decoder_only/vision_language/test_gguf_multimodal.py`

Performance (Required):
4. Text-only latency ≤ FlashAttention-2 latency (no regression)
5. Multimodal latency < TritonAttention latency (improvement amount TBD during implementation)
6. Memory usage for masks: O(N) verified via `torch.cuda.memory_allocated()`

Integration (Required):
7. Compatible with PagedAttention block tables
8. Compatible with CUDA graph capture (PIECEWISE mode)
9. Fallback to TritonAttention on unsupported GPUs (SM < 8.0)

Tests to Add:
- `tests/attention/test_flashmask_correctness.py`
- `benchmarks/attention/benchmark_flashmask.py`

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Isotr0py** [comment]
cc @LucasWilkinson about attention backend and FA.

**Comment by lucianommartins** [comment]
Hi @LucasWilkinson thanks for the feedbacks!

1. I think FA3 CuTE DSL FlexAttention is a great idea, but as FA3 in general, it is supported on Hopper platform (no SM < 9) and looks like the FA3 implementation for GB200 (SM 10) is not ready too. So I'm targetting FlashMask as a performance enhancement with the most platform coverage possible - Although I agree it would be a great new RFC to have FA3 CuTE DSL FlexAttention too!

2. I'm working with @Isotr0py to have #30386 merged and officially have TritonAttention implemented... it is not merged yet

3. Sure, I can remove the backward pass support. It can be added in the future if needed.

Thanks again!

**Comment by LucasWilkinson** [comment]
Seems like this only requires a few lines of change in FA2? if this is the case we should consider vendoring it via https://github.com/vllm-project/flash-attention instead of directly in `csrc`

**Comment by lucianommartins** [comment]
I still didn't go deep enough on the code implementation plan @LucasWilkinson - for now I'm not sure if I can simply reusing the FA2 CUDA kernels and simply do the change via python code changes.

But if it is feasible, I agree and certainly go through that way. But I feel that as the mask representation/storing mechanism is different, I will need to have the CUDA kernel ops aware of that and working correctly with that as well.
